### PR TITLE
Initial build

### PR DESCRIPTION
**Pull request by Zemurin**
- no math. Yes the rest.

### CODE REVIEW

File: GOLExtractor/Source/NamedColors/NamedColors.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+{
+	registerKeys();
+	parseFile(filepath);
+	clearRegisteredKeywords();
+}
+
+void NamedColors::registerKeys()
+{
+	// If we get a color named "colors", we're f--d.
+	registerKeyword("colors", [](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/NamedColors/NamedColors.cpp]:
```suggestion
	registerKeyword("colors", [](std::istream& theStream) {
```

File: GOLExtractor/Source/FlagFoundry/Warehouse.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class CoatOfArms;
+class Emblem;
+struct coloredImage
+{
+	std::optional<commonItems::Color> color1;
+	std::optional<commonItems::Color> color2;
+	std::optional<commonItems::Color> color3;
+	std::string patternName;
+	Magick::Image imageData;
+	bool operator==(const coloredImage& rhs);
```
Comment (IhateTrains) [inline GOLExtractor/Source/FlagFoundry/Warehouse.h:18]: Did you try defaulting it?

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	registerKeyword("sub", [this](const std::string& unused, std::istream& theStream) {
+		subs.emplace_back(std::make_shared<CoatOfArms>(theStream, ""));
+	});
+	registerKeyword("instance", [this](const std::string& unused, std::istream& theStream) {
+		auto instance = EmblemInstance(theStream);
+		if (instance.getOffset().empty())
+			instance.defaultOffset();
+		instances.emplace_back(instance);
+	});
+	registerKeyword("parent", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("parent", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+				suspiciousItem = result;
+			}
+		}
+		auto coaStream = std::stringstream(suspiciousItem);
+		coloredEmblems.emplace_back(Emblem(coaStream, color1, color2, color3));
+	});
+	registerKeyword("sub", [this](const std::string& unused, std::istream& theStream) {
+		subs.emplace_back(std::make_shared<CoatOfArms>(theStream, ""));
+	});
+	registerKeyword("instance", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("instance", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			{
+				std::string result;
+				std::regex_replace(std::back_inserter(result), suspiciousItem.begin(), suspiciousItem.end(), std::regex(macroID), macroValue);
+				suspiciousItem = result;
+			}
+		}
+		auto coaStream = std::stringstream(suspiciousItem);
+		coloredEmblems.emplace_back(Emblem(coaStream, color1, color2, color3));
+	});
+	registerKeyword("sub", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("sub", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color3 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("textured_emblem", [this](const std::string& unused, std::istream& theStream) {
+		texturedEmblems.emplace_back(Emblem(theStream, color1, color2, color3));
+	});
+	registerKeyword("colored_emblem", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("colored_emblem", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		{
+			color3 = laFabricaDeColor.getColor(theStream);
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color3 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("textured_emblem", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("textured_emblem", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		{
+			color2 = laFabricaDeColor.getColor(theStream);
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color2 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("color3", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("color3", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		{
+			color1 = laFabricaDeColor.getColor(theStream);
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color1 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("color2", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("color2", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+void CoatOfArms::registerKeys()
+{
+	registerRegex(R"(@[A-Za-z0-9_\-\']+)", [this](const std::string& macroID, std::istream& theStream) {
+		macros.insert(std::pair(macroID, commonItems::getString(theStream)));
+		});
+	registerKeyword("pattern", [this](const std::string& unused, std::istream& theStream) {
+		pattern = commonItems::singleString(theStream).getString();
+	});
+	registerKeyword("color1", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("color1", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+	parseStream(theStream);
+	clearRegisteredKeywords();
+}
+
+void CoatOfArms::registerKeys()
+{
+	registerRegex(R"(@[A-Za-z0-9_\-\']+)", [this](const std::string& macroID, std::istream& theStream) {
+		macros.insert(std::pair(macroID, commonItems::getString(theStream)));
+		});
+	registerKeyword("pattern", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
	registerKeyword("pattern", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+{
+	registerKeys();
+	parseStream(theStream);
+	clearRegisteredKeywords();
+}
+
+void CoatOfArms::registerKeys()
+{
+	registerRegex(R"(@[A-Za-z0-9_\-\']+)", [this](const std::string& macroID, std::istream& theStream) {
+		macros.insert(std::pair(macroID, commonItems::getString(theStream)));
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatOfArms.cpp]:
```suggestion
		macros.emplace(macroID, commonItems::getString(theStream));
```

File: GOLExtractor/Source/CoatsOfArms/CoatsOfArms.h [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+#define COATSOFARMS_H
+#include "Parser.h"
+
+class CoatOfArms;
+class CoatsOfArms: commonItems::parser
+{
+  public:
+	CoatsOfArms() = default;
+	void loadCoats(const std::string& filepath);
+	[[nodiscard]] auto getCoats() const { return coats; }
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatsOfArms.h]:
```suggestion
	[[nodiscard]] const auto& getCoats() const { return coats; }
```

File: GOLExtractor/Source/CoatsOfArms/CoatsOfArms.cpp [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+{
+	registerKeys();
+	parseFile(filePath);
+	clearRegisteredKeywords();
+}
+
+void CoatsOfArms::registerKeys()
+{
+	registerRegex(R"(@[A-Za-z0-9_\-\']+)", [this](const std::string& macroID, std::istream& theStream) {		
+		macros.insert(std::pair(macroID, commonItems::getString(theStream)));
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/CoatsOfArms.cpp]:
```suggestion
		macros.emplace(macroID, commonItems::getString(theStream));
```

File: GOLExtractor/Source/CoatsOfArms/Emblem.cpp [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+	parseStream(theStream);
+	clearRegisteredKeywords();
+}
+
+void Emblem::registerKeys()
+{
+	registerRegex(R"(@[A-Za-z0-9_\-\']+)", [this](const std::string& macroID, std::istream& theStream) {
+		macros.insert(std::pair(macroID, commonItems::getString(theStream)));
+		});
+	registerKeyword("texture", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/Emblem.cpp]:
```suggestion
	registerKeyword("texture", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/Emblem.cpp [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+		texture = commonItems::singleString(theStream).getString();
+	});
+	registerKeyword("color1", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/Emblem.cpp]:
```suggestion
	registerKeyword("color1", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/Emblem.cpp [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+		{
+			color1 = laFabricaDeColor.getColor(colorStream);
+		}
+		catch (std::exception& e)
+		{			
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color1 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("color2", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/Emblem.cpp]:
```suggestion
	registerKeyword("color2", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/Emblem.cpp [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+		{
+			color2 = laFabricaDeColor.getColor(colorStream);
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color2 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("color3", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/Emblem.cpp]:
```suggestion
	registerKeyword("color3", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/Emblem.cpp [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+		{
+			color3 = laFabricaDeColor.getColor(colorStream);
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << e.what() << " - sidestepping with black.";
+			color3 = commonItems::Color(std::array<int, 3>{0, 0, 0});
+		}
+	});
+	registerKeyword("mask", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/Emblem.cpp]:
```suggestion
	registerKeyword("mask", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/Emblem.cpp [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+		mask = commonItems::intList(theStream).getInts();
+	});
+	registerKeyword("instance", [this](const std::string& unused, std::istream& theStream) {
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/Emblem.cpp]:
```suggestion
	registerKeyword("instance", [this](std::istream& theStream) {
```

File: GOLExtractor/Source/CoatsOfArms/EmblemInstance.cpp [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+	});
+	registerKeyword("position", [this](const std::string& unused, std::istream& theStream) {
+		position = commonItems::doubleList(theStream).getDoubles();
+	});
+	registerKeyword("scale", [this](const std::string& unused, std::istream& theStream) {
+		scale = commonItems::doubleList(theStream).getDoubles();
+	});
+	registerKeyword("offset", [this](const std::string& unused, std::istream& theStream) {
+		offset = commonItems::doubleList(theStream).getDoubles();
+	});
```
Comment (IhateTrains) [inline GOLExtractor/Source/CoatsOfArms/EmblemInstance.cpp:29]: You can remove unused here.

File: GOLExtractor/Source/FlagFoundry/FlagCrafter.cpp [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+
+std::pair<Magick::Image, Magick::Image> FlagCrafter::processEmblemsOnImage(const std::pair<Magick::Image, Magick::Image>& imagePair,
+	 const std::vector<std::pair<Emblem, Magick::Image>>& emblems) const
+{
+	if (emblems.empty())
+		return imagePair;
+
+	auto workingImage = imagePair;
+
+	for (const auto& emblemPair: emblems)
```
Comment (IhateTrains) [inline GOLExtractor/Source/FlagFoundry/FlagCrafter.cpp]: `for (const auto& [first, second]: emblems)`

File: GOLExtractor/Source/FlagFoundry/FlagCrafter.cpp [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+
+std::pair<Magick::Image, Magick::Image> FlagCrafter::processSubsOnImage(const std::pair<Magick::Image, Magick::Image>& imagePair,
+	 const std::vector<std::pair<CoatOfArms, Magick::Image>>& subs) const
+{
+	if (subs.empty())
+		return imagePair;
+
+	auto workingImagePair = imagePair;
+
+	for (const auto& subPair: subs)
```
Comment (IhateTrains) [inline GOLExtractor/Source/FlagFoundry/FlagCrafter.cpp]: `for (const auto& [first, second]: subs)`

File: GOLExtractor/Source/FlagFoundry/FlagFoundry.cpp [hunk 24]
```diff
@@ -0,0 +1,47 @@
+﻿#include "FlagFoundry.h"
+#include "../CoatsOfArms/CoatOfArms.h"
+#include "Magick++.h"
+#include "OSCompatibilityLayer.h"
+#include "Warehouse.h"
+#include <ranges>
+
+#include "../CoatsOfArms/CoatsOfArms.h"
+
+// This is the only class that interacts with imageMagick, outside of EU4World, which depends on this one.
```
Comment (IhateTrains) [inline GOLExtractor/Source/FlagFoundry/FlagFoundry.cpp]: May be a leftover comment.

File: GOLExtractor/Source/FlagFoundry/Warehouse.cpp [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+					toReturn.emplace_back(std::pair(emblem, storedImage.imageData)); // We have a match of patternName and all possible colors.
+					matchedEmblem = true;
+					break;
+				}
+			}
+		if (matchedEmblem)
+			continue; // We're done here.
+
+		// Do we have a base texture in storage from which we can craft an image?
+		if (!baseTextures.count(newImageBlock.patternName))
```
Comment (IhateTrains) [inline GOLExtractor/Source/FlagFoundry/Warehouse.cpp]:
```suggestion
		if (!baseTextures.contains(newImageBlock.patternName))
```

File: GOLExtractor/Source/FlagFoundry/Warehouse.h (hunk 2 above)
Comment (Zemurin) [reply GOLExtractor/Source/FlagFoundry/Warehouse.h:18]: No, the colors are pointers so I must compare whatever's behind them.

File: GOLExtractor/Source/Configuration.cpp [hunk 26]
```diff
@@ -0,0 +1,3 @@
+#include "Configuration.h"
+#include "Color.h"
+auto laFabricaDeColor = commonItems::Color::Factory();
```
Comment (Idhrendur) [inline GOLExtractor/Source/Configuration.cpp]: If the only point of Configuration is a place for the definition of this factory, I'd move the definition elsewhere (main.cpp even!) and remove Configuration

**Review by Idhrendur — APPROVED** [review]

**Pull request merged**